index.bs

```diff
@@ -4648,71 +4659,152 @@ The <dfn export for=commands>browsingContext.setViewport</dfn> command modifies
+   1. If |navigable| is not a [=/top-level traversable=], return [=error=] with
+      [=error code=] [=invalid argument=].
 
-   1. Let |device pixel ratio| be the |command
-      parameters|["<code>devicePixelRatio</code>"].
+   1. [=set/Append=] |navigable| to |navigables|.
 
-   1. For the |navigable| and all [=descendant navigables=]:
+1. Otherwise, if the <code>userContexts</code> field of |command parameters| is present:
 
-      1. Let |navigable| be the [=/navigable=] whose [=navigable/active document=] is
-         |navigable|'s [=navigable/active document=].
+   1. Let |user contexts| be a [=/set=].
 
-      1. If |device pixel ratio| is not null:
+   1. For each |user context id| of |command parameters|["<code>userContexts</code>"]:
 
-         1. When the [=select an image source from a source set=] are run, act as if
-            the implementation's pixel density was set to |device pixel ratio| when selecting an image.
+      1. Set |user context| to [=get user context=] with |user context id|.
 
-         1. For the purposes of the [=resolution media feature=], act as if
-            the implementation's resolution is |device pixel ratio| dppx scaled by the page zoom.
+      1. If |user context| is null, return [=error=] with [=error code=] [=no such user context=].
 
-         1. [=map/Set=] [=device pixel ratio overrides=][|navigable|] to |device pixel ratio|.
+      1. [=set/Append=] |user context| to |user contexts|.
 
-            Note: This will take an effect because of the patch of [[#patchs-determine-the-device-pixel-ratio]].
+   1. For each |user context| of |user contexts|:
 
-      1. Otherwise:
+      1. [=map/Set=] [=viewport overrides map=][|user context|] to a new [=/map=].
 
-         1. When the [=select an image source from a source set=] steps are run, use the implementation's default behavior,
-            without any changes made by previous invocations of these steps.
+      1. If |command parameters| [=map/contains=] the <code>viewport</code> field:
+
+         1. Set |viewport overrides map|[|user context|]["<code>viewport</code>"]
+            to |command parameters|["<code>viewport</code>"].
+
+      1. If |command parameters| [=map/contains=] the <code>devicePixelRatio</code> field:
+
+         1. Set |viewport overrides map|[|user context|]["<code>devicePixelRatio</code>"]
+            to |command parameters|["<code>devicePixelRatio</code>"].
+
+      1. [=list/For each=] |top-level traversable| in the list of all [=/top-level traversables=]
+         whose [=associated user context=] is |user context|:
+
+         1. [=list/Append=] |top-level traversable| to |navigables|.
+
+1. If |command parameters| [=map/contains=] the <code>viewport</code> field:
+
+   1. Let |viewport| be the |command parameters|["<code>viewport</code>"].
+
+   1. For each |navigable| of |navigables|:
+
+      1. [=Set viewport=] with |navigable| and |viewport|.
+
+1. Run the [[cssom-view-1#resizing-viewports]] steps.
```

Comment:
(not related to your PR, but those steps seem to require a document to run against?)